### PR TITLE
Convert error messages to use event recorder

### PR DESCRIPTION
**Pull request by wccsama**
**What type of PR is this?**
```
/kind cleanup
```
**What this PR does / why we need it**:
```
todo: Convert error messages to use event recorder.
```
**Which issue(s) this PR fixes**:

```
NONE
```
**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
NONE
```

### CODE REVIEW

File: pkg/controller/service/service_controller.go [hunk 1]
```diff
@@ -559,20 +559,21 @@ func getPortsForLB(service *v1.Service) ([]*v1.ServicePort, error) {
 		if protocol == "" {
 			protocol = sp.Protocol
 		} else if protocol != sp.Protocol && wantsLoadBalancer(service) {
-			// TODO:  Convert error messages to use event recorder
-			return nil, fmt.Errorf("mixed protocol external load balancers are not supported")
+			s.eventRecorder.Eventf(service, v1.EventTypeWarning, "mixed protocol external load balancers are not supported",
```
Comment (MrHohn) [inline pkg/controller/service/service_controller.go]: This is a strange place to do validation like this (this function is only triggered when comparing an updated service object). Also, on the above it says the cloud provider itself is responsible for the protocol validation. Why don't we remove the validation here and delegate `EnsureLoadBalancer()` for doing the check of mixed protocol?

Comment (wccsama) [reply pkg/controller/service/service_controller.go]: Yeah~, mixed protocol may be checked in the function EnsureLoadBalancer().

**Comment by MrHohn** [comment]
> Should I remove the validation, or just print err message like this pr.

@wccsama I would suggest removing the check completely so that `getPortsForLB()` is sorely used for getting the ports from service. The mixed protocol config is already gated by the validation on apiserver anyway:
https://github.com/kubernetes/kubernetes/blob/36981002246682ed7dc4de54ccc2a96c1a0cbbdb/pkg/apis/core/validation/validation.go#L3998-L4012

**Comment by wccsama** [comment]
> > Should I remove the validation, or just print err message like this pr.
> 
> @wccsama I would suggest removing the check completely so that `getPortsForLB()` is sorely used for getting the ports from service. The mixed protocol config is already gated by the validation on apiserver anyway:
> https://github.com/kubernetes/kubernetes/blob/36981002246682ed7dc4de54ccc2a96c1a0cbbdb/pkg/apis/core/validation/validation.go#L3998-L4012

@MrHohn Done, PTAL thanks~

File: pkg/controller/service/service_controller.go [hunk 2]
```diff
@@ -548,31 +548,22 @@ func (s *ServiceController) needsUpdate(oldService *v1.Service, newService *v1.S
...
-		}
 	}
-	return ports, nil
+	return ports
 }
 
 func portsEqualForLB(x, y *v1.Service) bool {
-	xPorts, err := getPortsForLB(x)
-	if err != nil {
+	xPorts := getPortsForLB(x)
```
Comment (MrHohn) [inline pkg/controller/service/service_controller.go]: Seems like we don't need to check nil?
```golang
	xPorts := getPortsForLB(x)
	yPorts := getPortsForLB(y)
	return portSlicesEqualForLB(xPorts, yPorts)
```

**Review by MrHohn — APPROVED** [review]
/approve

File: pkg/controller/service/service_controller.go (hunk 2 above)
Comment (wccsama) [reply pkg/controller/service/service_controller.go]: yeah~

**Comment by MrHohn** [comment]
This LGTM, can you squash the fix-up commits?

**Comment by wccsama** [comment]
@MrHohn  already fixed~

**Review by MrHohn — APPROVED** [review]
Thanks!
/lgtm

**Pull request merged**